pkg/trace/api/evp_intake_proxy.go

```diff
@@ -0,0 +1,252 @@
+	}
+
+	return &httputil.ReverseProxy{
+		Director:  director,
+		ErrorLog:  logger,
+		Transport: &evpProxyTransport{transport, endpoints, conf.EVPProxy.MaxPayloadSize},
+	}
+}
+
+// evpProxyTransport sends HTTPS requests to multiple targets using an
+// underlying http.RoundTripper. API keys are set separately for each target.
+// When multiple endpoints are in use the response from the first endpoint
+// is proxied back to the client, while for all aditional endpoints the
+// response is discarded.
+type evpProxyTransport struct {
+	transport      http.RoundTripper
+	endpoints      []config.Endpoint
+	maxPayloadSize int64
+}
+
+func (t *evpProxyTransport) RoundTrip(req *http.Request) (rresp *http.Response, rerr error) {
+
+	// Metrics with stats for debugging
+	beginTime := time.Now()
+	metricTags := []string{}
+	var contentLength int64
+	defer func() {
+		metrics.Count("datadog.trace_agent.evp_proxy.request", 1, metricTags, 1)
+		metrics.Count("datadog.trace_agent.evp_proxy.request_bytes", contentLength, metricTags, 1)
+		metrics.Timing("datadog.trace_agent.evp_proxy.request_duration_ms", time.Since(beginTime), metricTags, 1)
+		if rerr != nil {
+			metrics.Count("datadog.trace_agent.evp_proxy.request_error", 1, metricTags, 1)
+		}
+	}()
+
+	if req == nil || req.URL == nil {
+		return nil, fmt.Errorf("EVPProxy: invalid request")
+	}
+
+	if req.Body != nil && t.maxPayloadSize > 0 {
+		req.Body = apiutil.NewLimitedReader(req.Body, t.maxPayloadSize)
+	}
+
+	if ct := req.Header.Get("Content-Type"); ct != "" {
+		metricTags = append(metricTags, "content_type:"+ct)
+	}
+	contentLength = req.ContentLength
+
+	// Parse request path: The first component is the target subdomain, the rest is the target path.
+	inputPath := req.URL.Path
+	parts := strings.SplitN(inputPath, "/", 3)
+	if len(parts) != 3 || parts[0] != "" || parts[1] == "" || parts[2] == "" {
+		return nil, fmt.Errorf("EVPProxy: invalid path: '%s'", inputPath)
+	}
+	subdomain := parts[1]
+	req.URL.Path = strings.TrimPrefix(req.URL.Path, "/"+subdomain)
+
+	// Sanitize the input
+	if !isValidSubdomain(subdomain) {
+		return nil, fmt.Errorf("EVPProxy: invalid subdomain: %s", subdomain)
+	}
+	metricTags = append(metricTags, "subdomain:"+subdomain)
+	if !isValidPath(req.URL.Path) {
+		return nil, fmt.Errorf("EVPProxy: invalid target path: %s", req.URL.Path)
+	}
+	if !isValidQueryString(req.URL.RawQuery) {
+		return nil, fmt.Errorf("EVPProxy: invalid query string: %s", req.URL.RawQuery)
+	}
+
+	req.URL.Scheme = "https"
+	setTarget := func(r *http.Request, host, apiKey string) {
+		targetHost := subdomain + ".evp." + host
+		r.Host = targetHost
+		r.URL.Host = targetHost
+		r.Header.Set("DD-API-KEY", apiKey)
+	}
+
+	if len(t.endpoints) == 1 {
+		setTarget(req, t.endpoints[0].Host, t.endpoints[0].APIKey)
+		return t.transport.RoundTrip(req)
+	}
+
+	// There's more than one destination endpoint
+
+	var slurp *[]byte
+	if req.Body != nil {
```

Comment:
This overly defensive attitude doesn't really help us to be honest. When can `req.Body` be `nil` really? Maybe I'm unaware of some situation.